src/allmydata/storage_client.py

```diff
@@ -1189,3 +1190,53 @@ def advise_corrupt_share(
             )
         else:
             raise NotImplementedError()  # future tickets
+
+    @defer.inlineCallbacks
+    def slot_readv(self, storage_index, shares, readv):
+        mutable_client = StorageClientMutables(self._http_client)
+        reads = {}
+        # TODO if shares list is empty, that means list all shares, so we need
+        # to do a query to get that.
+        assert shares  # TODO replace with call to list shares
```

Comment:
I wonder if the higher level client code wants to read from all shares very often.  If it does, it would be nice (now or later) to not force an extra round-trip to the server to ask it what shares it has for that case.  If it doesn't, maybe we don't even need to support this case in the API/protocol.

Reply:
Let's see...

1. It's part of `IStorageClient` ("An empty shares list means to return data from all known shares., and I wrote a test for it there...") and so I have a interface-test for that in `test_istorageserver.py`.
2. `layout.py` passes in a single share number to `slot_readv`, so it always passing something in.
3. But `servermap.py` actually has an explicit path where it passes in an empty list :(

So it looks like this is used, yes.

Reply:
So two options:

1. Do a list shares operation whenever the list is empty.
2. Extend the protocol to support the "send all shares" case. Using empty list for that seems like bad semantics, but the protocol could be extended in a nicer way.

I tend towards option 2. What do you think?

Reply:
Yea I prefer (2) over (1).